src/types/mod.rs

```diff
@@ -424,6 +426,128 @@ impl ValueInternal for JsNumber {
     }
 }
 
+/// A JavaScript Date object.
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+#[cfg(feature = "napi-5")]
+pub struct JsDate(raw::Local);
+
+#[cfg(feature = "napi-5")]
+impl Value for JsDate { }
+
+#[cfg(feature = "napi-5")]
+impl Managed for JsDate {
+    fn to_raw(self) -> raw::Local { self.0 }
+
+    fn from_raw(_: Env, h: raw::Local) -> Self { JsDate(h) }
+}
+
+#[cfg(feature = "napi-5")]
+#[derive(PartialEq, PartialOrd, Clone, Debug)]
+pub struct DateError(f64, DateErrorKind);
+
+#[cfg(feature = "napi-5")]
+impl DateError {
+    pub fn kind(&self) -> DateErrorKind {
+        self.1
+    }
+}
+
+#[cfg(feature = "napi-5")]
+impl fmt::Display for DateError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.1.as_str())
+    }
+}
+
+#[cfg(feature = "napi-5")]
+impl Error for DateError {}
+
+#[cfg(feature = "napi-5")]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub enum DateErrorKind {
+    Overflow,
+    Underflow,
+}
+
+#[cfg(feature = "napi-5")]
+impl DateErrorKind {
+    pub fn as_str(&self) -> &'static str {
+        match *self {
+            DateErrorKind::Overflow => "date overflow",
+            DateErrorKind::Underflow => "date underflow",
+        }
+    }
+}
+
+#[cfg(feature = "napi-5")]
+pub type DateResult<'a> = Result<Handle<'a, JsDate>, DateError>;
+
+#[cfg(feature = "napi-5")]
+impl<'a> JsResultExt<'a, JsDate> for DateResult<'a> {
+    /// Creates a `Error` on error
+    fn or_throw<'b, C: Context<'b>>(self, cx: &mut C) -> JsResult<'a, JsDate> {
+        match self {
+            Ok(v) => Ok(v),
+            Err(e) => cx.throw_error(&e.to_string())
+        }
+    }
+}
+
+#[cfg(feature = "napi-5")]
+impl JsDate {
+    pub const MIN_VALUE: f64 = -8.64e15;
+    pub const MAX_VALUE: f64 = 8.64e15;
+
+    pub fn new<'a, C: Context<'a>, T: Into<f64>>(cx: &mut C, time: T) -> Handle<'a, JsDate> {
+        let env = cx.env().to_raw();
+        let local = unsafe {
+            neon_runtime::date::new_date(env, time.into())
+        };
+        Handle::new_internal(JsDate(local))
+    }
+
+    pub fn new_lossy<'a, C: Context<'a>, V: Into<f64> + std::cmp::PartialOrd>(cx: &mut C, value: V) -> DateResult<'a> {
+        let env = cx.env().to_raw();
+        let time = value.into();
+        let local = unsafe {
+            neon_runtime::date::new_date(env, time.clone())
+        };
+        let date = Handle::new_internal(JsDate(local));
+        if date.is_valid(cx) { return Ok(date); }
+
+        if time > JsDate::MAX_VALUE {
```

Comment:
Can we move this to the beginning of the function right after `let time = value.into();`? It saves the construction of the date object if we're going to throw it away.